DFE-Digital/plan-technology-for-your-school
# PoC: Automated process for fetching gias data

## Overview

This PR is a proof of concept for the ticket [#236144](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/236144)

**Requires reverting the TODO comments before merge - if this does get merged**

It includes:
- a python project to fetch GIAS data and store it in our SQL database
- a workflow to run this daily
- changes to the web app to store UID on login

The pipeline failures are just because the SQL script hasn't been run on dev yet
`TODO` comments are in place, anywhere a code change is specifically for the PoC and should be reverted.
This could be merged in as is with the TODOs reverted, however The C# part of this may be subject to change depending on whats required

## Changes

### Python project for Updating GIAS data
- GIAS data is refreshed daily and dynamically generated on request, so this uses playwright to get the relevant files
- Uses pandas and pyodbc to save the data into tables in our plan tech sql db
- In future, this might be replaced with a direct connection to the databricks instance that has the data used to generate the GIAS files avoiding the need to run this daily

### Pipeline
- Workflow to fetch and save the GIAS data on a schedule, similar to other ones
- Currently requires permission to run on prod - I suggest making another environment for automated jobs that require prod secrets so that it doesn't have to be approved by a developer every day
- With a direct connection to databricks, this pipeline could be removed too

### Update web app
- Stores the UID from the claims in the web app
- Looking up establishments on login is then just a case of 
```sql
Select EL.establishmentName
from dbo.establishment E
Join dbo.establishmentGroup EG on E.groupUid = EG.uid
Join dbo.establishmentLink EL on EL.groupUid = EG.uid
Where @establishmentId = E.Id
```

## How to review the PR
- Use the Seed test database project to make a local database with the newest db script run on it
- Setup the python project locally, ensuring the README provides enough information to successfully get it running
- run the python project against the seeded test db instead of the real one
- see that the data is downloaded, extracted and the database updated
- On login, all establishments connected to the simulated one (see one of the TODO comments) should show on the self assessment page

Alternatively run the SQL script on dev, and use the dev db
and after running on dev
- Check that the pipeline which is running on pull_request at the moment is functioning correctly

## Release requirements
- Add the python connection string to the keyvault for test/staging/prod

## Checklist

- [ ] revert the `pull_request` condition on the workflow, it should only be cron & dispatch - pr is just for testing purposes
